lib/twom.c

```diff
@@ -0,0 +1,3293 @@
+            return TWOM_IOERROR;
+        }
+        break;
+    }
+
+    int r = twom_db_begin_txn(db, TWOM_SHARED|TWOM_MVCC, &txn);
+    if (r) return r;
+
+    if (txn->file->fd != oldfd) {
+        // we lost a race here! When the transaction locked the file,
+        // it got a new inode, so another process has repacked meanwhile.
+        r = TWOM_LOCKED;
+        goto badfile;
+    }
+
+    // don't make things worse with a broken file
+    r = twom_txn_consistent(txn);
+    if (r) {
+        db->error("inconsistent pre-repack",
+                  "filename=<%s>", db->fname);
+        goto badfile;
+    }
+
+    /* open fname.NEW */
+    snprintf(newfname, sizeof(newfname), "%s.NEW", db->fname);
+
+    int flags = TWOM_NONBLOCKING; // if another file has already locked it, something is broken
+    if (db->external_csum)
+        flags |= TWOM_CSUM_EXTERNAL;
+    else if (db->nocsum)
+        flags |= TWOM_CSUM_NULL;
+    else
+        flags |= TWOM_CSUM_XXH64;
+    if (db->external_compar)
+        flags |= TWOM_COMPAR_EXTERNAL;
+    if (db->nosync)
+        flags |= TWOM_NOSYNC;
+
+    /* we hand-open the new file and make it the first openfile, so it has
+     * the write_txn against it and all the sanity checks make sense */
+    unlink(newfname);
+    newfd = open(newfname, O_RDWR|O_CREAT, 0644);
+    if (newfd < 0) {
+        r = TWOM_IOERROR;
+        goto badfile;
+    }
+
+    struct tm_file *oldfile = db->openfile;
+    db->openfile = (struct tm_file *)twom_zmalloc(sizeof(struct tm_file));
+    db->openfile->fd = newfd;
+    db->openfile->next = oldfile;
+
+    // we're just doing small copies, release less frequently
+    db->foreach_lock_release *= 64;
+
+    // write lock will create a new header if the file is zero sized.
+    // It will have a different UUID, no records, and be generation 1
+    r = write_lock(db, &newtxn, db->openfile, flags);
+    if (r) goto fail;
+
+    // we'll likely need about enough space for the current
+    // database, minus the dirty bytes, minus 24 bytes per
+    // extra commit.  This isn't exact, because new records
+    // will have diferent random levels
+    tm_ensure(db, oldfile->header.current_size
+                - oldfile->header.dirty_size
+                - (oldfile->header.num_commits-1) * 24);
+
+    // initialise the writer location on the new file
+    r = find_loc(newtxn, &db->loc, NULL, 0);
+    if (r) goto fail;
+
+    // mvcc process all the existing records
+    r = twom_txn_foreach(txn, NULL, 0, NULL, copy_cb, newtxn, /*flags*/0);
+    if (r) goto fail;
+
+    /* remember the repack size at this point, because that's everything
+     * which is in order!  The remaing replayed items might be deletes or
+     * replaces, so they still dirty up the ordering as much as new changes. */
+    newtxn->file->header.repack_size = newtxn->end;
+
+    // replay all the remaining changes to the end of the file
+    r = myreplay(txn, replay_cb, newtxn);
```

Comment:
The value stored here in `r` is never read.

Reply:
That was a real bug.  It's fixed with an `if (r) goto fail;`